src/Nethermind/Nethermind.DataMarketplace.Consumers/Services/ConsumerService.cs

```diff
@@ -122,18 +125,77 @@ public class ConsumerService : IConsumerService
             _receiptRequestValidator = receiptRequestValidator;
             _refundService = refundService;
             _blockchainBridge = blockchainBridge;
+            _blockProcessor = blockProcessor;
             _consumerAddress = consumerAddress ?? Address.Zero;
             _nodePublicKey = nodePublicKey;
             _timestamper = timestamper;
             _consumerNotifier = consumerNotifier;
-            _depositRequiredConfirmations = depositRequiredConfirmations;
+            _requiredBlockConfirmations = requiredBlockConfirmations;
             _logger = logManager.GetClassLogger();
-            _timer = new Timer {Interval = 5000};
-            _timer.Elapsed += OnTimeElapsed;
-            _timer.Start();
             _wallet.AccountLocked += OnAccountLocked;
             _wallet.AccountUnlocked += OnAccountUnlocked;
             _accountLocked = !_wallet.IsUnlocked(_consumerAddress);
+            _blockProcessor.BlockProcessed += OnBlockProcessed;
+            _timer = new Timer(TryClaimRefundsIntervalSeconds * 1000);
+            _timer.Elapsed += TimerOnElapsed;
+            _timer.Start();
+        }
+
+        private void TimerOnElapsed(object sender, ElapsedEventArgs e)
+        {
+            if (_logger.IsInfo) _logger.Info("Verifying whether any refunds might be claimed...");
+            _depositRepository.BrowseAsync(new GetDeposits
+                {
+                    Results = int.MaxValue,
+                    EligibleToRefund = true,
+                    CurrentBlockTimestamp = _currentBlockTimestamp
+                })
+                .ContinueWith(async t =>
+                {
+                    if (t.IsFaulted && _logger.IsError)
+                    {
+                        _logger.Error($"Fetching the deposits has failed.", t.Exception);
+                        return;
+                    }
+
+                    await TryClaimRefundsAsync(t.Result.Items);
+                });
+        }
+
+        private void OnBlockProcessed(object sender, BlockProcessedEventArgs e)
+        {
+            _currentBlockTimestamp = (ulong) e.Block.Timestamp;
```

Comment:
is it atomic?

Reply:
Blocks are being processed one by one, so it should be, however, I'll use `Interlocked` to make sure that there are no concurrency issues whatsoever.